design/BufferOperations.md

```diff
@@ -0,0 +1,183 @@
+# Buffer operations
+
+This explainer describes the operations that are available on the `GPUBuffer` object directly.
+They are `setSubData` which is an immediate data upload operations, and `mapWriteAsync`, `mapReadAsync` and `unmap` which are memory mapping operations.
+
+## Prelimineries: buffered / unbuffered commands
+
+Assuming there is a single queue, there are two types of commands in WebGPU:
+
+ - "Buffered commands": any commands on a `GPUCommandBuffer`, `GPUComputePassEncoder` or `GPURenderPassEncoder`.
+ - "Unbuffered commands": all other commands.
+
+Assuming there is a single queue, there is a total order on the unbuffered commands: they all execute atomically in the order they were called.
+`GPUQueue.submit` is special because it atomically executes all the commands stored in its `commands` argument.
+
+## `GPUBuffer.setSubData`
+
+```webidl
+partial interface GPUBuffer {
+    void setSubData(u32 offset, ArrayBuffer data);
+}
+```
+
+Calling `setSubData` on a `buffer` causes its content in the range `[offset, offset + data.length)` to be updated with the content of `data`.
+The update to the content of `buffer` happens atomically after all previous unbuffered operations and before all subsequent unbuffered operations.
+
+The following must be true or a validation error occurs on the call to `setSubData`:
+
+ - `buffer` must have been created with the `TRANSFER_DST` usage flag.
+ - `offset + data.length` must be less or equal to the size of `buffer`.
+ - `buffer` must be in the unmapped state (which also means it hasn't been destroyed with `GPUBuffer.destroy`)
+ - `data.length` and `offset` aren't aligned (TODO investigate alignment constraints)
+
+## Buffer mapping
+
+### `MAP_READ` and `MAP_WRITE`
+
+The `MAP_READ` and `MAP_WRITE` buffer creation usage flags need to be specified to create a buffer mappable for reading (resp. for writing).
+An additional validation constraint is that the `MAP_READ` usage flag must not be used with any other flags except with `TRANSFER_DST`.
```

Comment:
I agree with @kvark , the flag constraints are too tight.

Reply:
This was meant as a validation constraint at buffer creation time and could mostly be lifted. I'd like to prevent `GPUBuffers` from being both `MAP_WRITE` and `MAP_READ` because my understanding is that on desktop GPUs the best memory type to use depends on upload vs. download use cases.

Reply:
But Read and Write buffers can help us save significant amount of memory.

I.e. right now I'm porting some code to use `std::vector` over an allocator that hands out a READ&WRITE mapping pointer to memory that lives entirely in the non-device local memory heap.

This actually saves me a memcpy.

Reply:
How would you deal with Vulkan devices that only expose a COHERENT | HOST_VISIBLE (not cached) memory type? It seems really bad for reads.

We can always lift this restriction later. Changed to say "MAP_READ and MAP_WRITE can't be used in combination"